cwl/src/main/scala/cwl/WorkflowStep.scala

```diff
@@ -281,36 +352,131 @@ case class WorkflowStep(
-          - accumulate all that in the WorkflowStepInputFold
-        2) Fold over the callable input definition using the expression node map from 1):
-          - determine the correct mapping for the input definition based on the expression node map
-          and the type of input definition
-          - accumulate those mappings, along with potentially newly created graph input nodes as well as call input ports
-          in an InputDefinitionFold
-        3) Use the InputDefinitionFold to build a new call node
+        * This (big) flatMap builds nodes from top to bottom in the diagram above.
+        * If necessary, the scatter node is built last as it wraps some of the other nodes.
        */
       for {
+        /* ************************************ */
+        /* ************ Merge Nodes *********** */
+        /* ************************************ */
+        // Build merge nodes and recursively generates other call nodes that we haven't seen so far
         stepInputFold <- stepInputFoldCheck
+        // Extract the merge nodes from the fold
+        mergeNodes = stepInputFold.mergeNodes
+
+        /* ************************************ */
+        /* ****** Scatter Variable Nodes ****** */
+        /* ************************************ */
+        scatterVariableNodes <- ScatterLogic.buildScatterVariableNodes(scatter, mergeNodes, unqualifiedStepId.localName.value)
+
+        /* ************************************ */
+        /* *************** OGINS ************** */
+        /* ************************************ */
+        ogins = buildOGINs(mergeNodes, scatterVariableNodes)
+
+        /* ************************************ */
+        /* ********* Expression Nodes ********* */
+        /* ************************************ */
+        // Aggregate the generated nodes so far. This map will be used to generate expression nodes, so the order of aggregation matters:
+        // scatter variables and ogins take precedence over merge nodes (see diagram)
+        expressionNodeMap = mergeNodes ++ scatterVariableNodes ++ ogins
+        // Build expression nodes for inputs that have a valueFrom field
+        stepInputExpressionNodes <- buildStepInputExpressionNodes(expressionNodeMap)
+
+        /* ************************************ */
+        /* ************* Call Node ************ */
+        /* ************************************ */
+        // Get the callable object for this step
         checkedCallable <- callable
-        scatterVariables <- ScatterLogic.buildScatterVariables(scatter, stepInputFold, unqualifiedStepId.localName.value)
-        inputDefinitionFold <- checkedCallable.inputs.foldMap(foldInputDefinition(stepInputFold.stepInputMapping, scatterVariables)).toEither
+        // Aggregate again by adding generated expression nodes. Again order matters here, expression nodes override other nodes.
+        pointerMap = expressionNodeMap.asIdentifierMap ++ stepInputExpressionNodes
```

Comment:
The variable naming is bit confusing here, everything seems to be expressiony. It seems that the workflow step inputs are trumping the other things (which were inputs to the workflow step input expressions)?